co2mond/src/main.c

```diff
@@ -122,33 +167,230 @@ static void
+                "Connection: close\r\n"
+                "\r\n"
+                "goto /metrics;\r\n"
+            );
+            goto flush;
+        }
+
+        state_lock();
+        memcpy(&copy, &co2mon, sizeof(copy));
+        state_unlock();
+
+        if (!bitarr_isset(copy.seen, CODE_TAMB) || !bitarr_isset(copy.seen, CODE_CNTR))
+        {
+            fprintf(out,
+                "HTTP/1.0 503 Service Unavailable\r\n"
+                "Server: co2mond\r\n"
+                "Connection: close\r\n"
+                "\r\n"
+                "Device not ready.\r\n"
+            );
+            goto flush;
+        }
+
+        fprintf(out,
+            "HTTP/1.0 200 OK\r\n"
+            "Server: co2mond\r\n"
+            "Connection: close\r\n"
+            "\r\n"
+        );
+
+        // Note, HTTP has \r\n and Prometheus uses \n as line separator.
+
+        if (print_unknown)
+        {
+            int has_unknown = 0;
+            for (int i = 0; i < sizeof(copy.data) / sizeof(copy.data[0]); ++i)
+            {
+                if (bitarr_isset(copy.seen, i) && i != CODE_TAMB && i != CODE_CNTR)
+                {
+                    has_unknown = 1;
+                    break;
+                }
+            }
+            if (has_unknown)
+            {
+                fprintf(out,
+                    "# HELP co2mon_unknown Unknown value.\n"
+                    "# TYPE co2mon_unknown gauge\n"
+                );
+                for (int i = 0; i < sizeof(copy.data) / sizeof(copy.data[0]); ++i)
+                {
+                    if (bitarr_isset(copy.seen, i) && i != CODE_TAMB && i != CODE_CNTR)
+                    {
+                        fprintf(out, "co2mon_unknown{key=\"x%02x\"} %d\n", i, copy.data[i]);
+                    }
+                }
+            }
+        }
+        fprintf(out,
+            "# HELP co2mon_temp_celsius Ambient temperature.\n"
+            "# TYPE co2mon_temp_celsius gauge\n"
+            "co2mon_temp_celsius %.4f\n"
+            "# HELP co2mon_co2_ppm Concentration of CO2, parts per million.\n"
+            "# TYPE co2mon_co2_ppm gauge\n"
+            "co2mon_co2_ppm %d\n"
+            "# HELP co2mon_device_errors_total CO2 monitor device error counter.\n"
+            "# TYPE co2mon_device_errors_total counter\n"
+            "co2mon_device_errors_total %d\n"
+            "# HELP co2mon_heartbeat_time_seconds CO2 monitor heartbeat timestamp.\n"
+            "# TYPE co2mon_heartbeat_time_seconds gauge\n"
+            "co2mon_heartbeat_time_seconds %lld\n",
+            decode_temperature(copy.data[CODE_TAMB]),
+            copy.data[CODE_CNTR],
+            copy.deverr,
+            (long long)copy.heatbeat
+        );
+flush:
+        fflush(out);
+        if (shutdown(client_fd, SHUT_WR) != 0)
+        {
+            perror("shutdown");
+            goto cleanup;
+        }
+        fgetc(out); // wait till EOF (or timeout) before calling close()
```

Comment:
TBH, I'm unsure if this `fgetc()` is actually useful as `close()` should send the queue the background unless SO_LINGER is explicitly set (according to `man 7 socket`).